### PR TITLE
Integrate with PlanResourceChange and ApplyResourceChange

### PR DESCRIPTION
**Pull request by t0yv0**
Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1505

Requires https://github.com/pulumi/terraform-plugin-sdk/pull/35

sdk-v2 bridge has a new option that changes the implementation of resource lifecycle to go through PlanResourceChange and ApplyResourceChange TF SDKv2 gRPC methods.

```go
// Selectively opt-in resources that pass the filter to using PlanResourceChange. Resources are
// identified by their TF type name such as aws_ssm_document.
func WithPlanResourceChange(filter func(tfResourceType string) bool) providerOption { //nolint:revive
```

Enables fixing:  https://github.com/pulumi/pulumi-aws/issues/2555

Known differences: state returned by the new method will include explicit entries `"foo": null` for properties that are known by the schema but not set in the state. This seems to be benign for repeated diffs and refresh applications.

### CODE REVIEW

**Comment by t0yv0** [comment]
ignoreChanges has a bit of a chicken-and-egg problem because of how tfbridge code is written, e.g. Update.


```go
	diff, err := p.tfc().DiffWithContext(ctx, res.TFName, state, config)
	if err != nil {
		return nil, errors.Wrapf(err, "diffing %s", urn)
	}
	if diff == nil {
		// It is very possible for us to get here with a nil diff: custom diffing behavior, etc. can cause
		// textual/structural changes not to be semantic changes. A better solution would be to change the result of
		// Diff to indicate no change, but that is a slightly riskier change that we'd rather not take at the current
		// moment.
		return &pulumirpc.UpdateResponse{Properties: req.GetOlds()}, nil
	}

	// Apply any ignoreChanges before we check that the diff doesn't require replacement or deletion since we may be
	// ignoring changes to the keys that would result in replacement/deletion.
	doIgnoreChanges(ctx, res.TF.Schema(), res.Schema.Fields, olds, news, req.GetIgnoreChanges(), diff)

	if diff.Destroy() || diff.RequiresNew() {
		return nil, fmt.Errorf("internal: expected diff to not require deletion or replacement"+
			" during Update of %s. Found delete=%t, replace=%t. This indicates a bug in provider.",
			urn, diff.Destroy(), diff.RequiresNew())
	}

	if req.Timeout != 0 {
		diff.SetTimeout(req.Timeout, shim.TimeoutUpdate)
	}
```

Currently PlanResourceChanges computes the terraform.InstanceDiff object AND computes the planned cty.Value that's the result of applying that diff; technically processing ignore changes (or timeouts) should modify BOTH, but this is awkward in the current API. Possibly we can refactor a bit so we push ignoreChanges processor to DiffWithContext. Or else switch this out entirely to use Pulumi-level ignoreChanges like for PF. Tricky.

**Comment by t0yv0** [comment]
I need to check in an additional patch to go-sdk, missed it with go.work.

**Comment by t0yv0** [comment]
https://github.com/pulumi/terraform-plugin-sdk/pull/34 the necessary change.

**Comment by VenelinMartinov** [comment]
This sounds pretty promising to me since it'd allow us to get rid of some of our custom behaviour there. LMK if you need any help to push this through.

File: pkg/tfshim/sdk-v2/provider2.go [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+	return &v, nil
+}
+
+func (p *v2Provider2) Apply(
+	t string, s shim.InstanceState, d shim.InstanceDiff,
+) (shim.InstanceState, error) {
+	return p.ApplyWithContext(context.Background(), t, s, d)
+}
+
+func (*v2Provider2) unpackDiff(ty cty.Type, d shim.InstanceDiff) *v2InstanceDiff2 {
```
Comment (iwahbe) [inline pkg/tfshim/sdk-v2/provider2.go:197]: I would love a comment explaining why this is necessary.

File: pkg/tfshim/sdk-v2/provider2.go [hunk 2]
```diff
@@ -0,0 +1,683 @@
...
+	var r *v2InstanceState2
+	if s == nil {
+		return &v2InstanceState2{
+			resourceType: t,
+			stateValue:   cty.NullVal(ty),
+		}
+	} else {
+		r = s.(*v2InstanceState2)
+	}
+	return r
```
Comment (iwahbe) [inline pkg/tfshim/sdk-v2/provider2.go:256]: Nit

```suggestion
	if s == nil {
		return &v2InstanceState2{
			resourceType: t,
			stateValue:   cty.NullVal(ty),
		}
	}
	return s.(*v2InstanceState2)
```

File: pkg/tfshim/sdk-v2/provider_options.go [hunk 3]
```diff
@@ -49,3 +50,11 @@ func getProviderOptions(opts []providerOption) (providerOptions, error) {
 	}
 	return res, nil
 }
+
+// Selectively opt-in resources that pass the filter to using PlanResourceChange.
+func WithPlanResourceChange(filter func(string) bool) providerOption {
```
Comment (iwahbe) [inline pkg/tfshim/sdk-v2/provider_options.go:55]: Does filter take a tf token or a pulumi token?

```suggestion
func WithPlanResourceChange(filter func(string) bool) providerOption {
```

**Review by iwahbe — COMMENTED** [review]
Sanity checking my understanding of the implementation, we introduce a new `pkg/tfshim` `Provider` called `v2Provider2`, which thinks in terms of `PlanResourceChange` and `ApplyResourceChange`. We also introduce a binary muxer that switches on resource name called `providerWithPlanResourceChange`.

I think the overall strategy is good. Do you have a sense of how much work is needed to ship for https://github.com/pulumi/pulumi-aws/issues/2555?

> Currently PlanResourceChanges computes the terraform.InstanceDiff object AND computes the planned cty.Value that's the result of applying that diff; technically processing ignore changes (or timeouts) should modify BOTH, but this is awkward in the current API. Possibly we can refactor a bit so we push ignoreChanges processor to DiffWithContext. Or else switch this out entirely to use Pulumi-level ignoreChanges like for PF. Tricky.

I'm not aware of any customer pushback on the Pulumi-level ignoreChanges, so my inclination is to do that. (for simplicity and consistency).

**Comment by t0yv0** [comment]
> Do you have a sense of how much work is needed to ship for https://github.com/pulumi/pulumi-aws/issues/2555?

Optimistically there's 2 days left to work out ignoreChanges and refresh concerns, after which we can enable it in pulumi-aws for this resource only.

**Comment by t0yv0** [comment]
Unfortunately, eh, the refresh and import support turn out to be quite coupled to the state representation, so they need to be part of the change. I'm adding some code to do something sensible but it adds to what needs to be tested out. 

Still on the list - this needs to do state upgrades still and do timeouts correctly.

**Comment by t0yv0** [comment]
> I'm not aware of any customer pushback on the Pulumi-level ignoreChanges, so my inclination is to do that. (for simplicity and consistency).

That's probably because the user base for PF-based providers is still insignificant. I'm retaining the current ignoreChanges model, we can contemplate a change as a separate work item.

File: pkg/tfbridge/provider_test.go [hunk 4]
```diff
@@ -248,13 +248,14 @@ func testIgnoreChanges(t *testing.T, provider *Provider) {
 	})
 	assert.NoError(t, err)
 
-	outs, err := plugin.UnmarshalProperties(createResp.GetProperties(), plugin.MarshalOptions{KeepUnknowns: true})
+	outs, err := plugin.UnmarshalProperties(createResp.GetProperties(), plugin.MarshalOptions{
+		KeepUnknowns: true,
+		SkipNulls:    true,
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:253]: This was interesting, without SkipNulls: true the PlanResourceChange version returns explicit null entries for every property in the schema that's unset.

File: pkg/tfbridge/provider_test.go [hunk 5]
```diff
@@ -248,13 +248,14 @@ func testIgnoreChanges(t *testing.T, provider *Provider) {
...
+	})
 	assert.NoError(t, err)
-	assert.True(t, resource.PropertyMap{
-		"id":                  resource.NewStringProperty(""),
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:254]: The PlanResourceChange version returns an unknown for "id", but that seems like may be OK.

File: pkg/tfbridge/provider_test.go [hunk 6]
```diff
@@ -248,13 +248,14 @@ func testIgnoreChanges(t *testing.T, provider *Provider) {
...
-		"stringPropertyValue": resource.NewStringProperty("foo"),
-		"setPropertyValues":   resource.NewArrayProperty([]resource.PropertyValue{resource.NewStringProperty("foo")}),
-	}.DeepEquals(outs))
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:257]: Also the PlanResourceChange version returns __meta here with resource timeouts populated. That seems to be a change.

File: pkg/tfbridge/provider_test.go [hunk 7]
```diff
@@ -323,17 +324,28 @@ func TestIgnoreChanges(t *testing.T) {
 }
 
 func TestIgnoreChangesV2(t *testing.T) {
+	testIgnoreChangesV2(t, shimv2.NewProvider(testTFProviderV2))
+}
+
+func TestIgnoreChangesV2WithPlanResourceChange(t *testing.T) {
+	opt := shimv2.WithPlanResourceChange(func(string) bool { return true })
+	testIgnoreChangesV2(t, shimv2.NewProvider(testTFProviderV2, opt))
```
Comment (t0yv0) [inline pkg/tfbridge/provider_test.go:332]: Starting to bring this option into the test matrix to get some coverage..

**Comment by t0yv0** [comment]
Testing import was fruitful, some tweaks needed to handle not-found resource. I'll have to look into adding coverage for refresh and import, found and not found.

**Comment by t0yv0** [comment]
Got most of the provider2 covered, Remaining coverage :

  - test emulating 2555 specifically
  - recoverDiagnostic; diagnostic validity
  - resource destroy
  - state upgrades
  - timeouts

Manual testing: testing 2555 scenario on top of the fixes; also testing import and refresh.

File: pkg/tfbridge/provider.go [hunk 8]
```diff
@@ -1329,7 +1329,7 @@ func (p *Provider) Delete(ctx context.Context, req *pulumirpc.DeleteRequest) (*p
 	}
 
 	// Create a new destroy diff.
-	diff := p.tf.NewDestroyDiff(ctx, string(t), shim.TimeoutOptions{
```
Comment (VenelinMartinov) [inline pkg/tfbridge/provider.go:1332]: This looks like a fairly big change? Isn't urn.Type() a pulumi type, while this requires a TF type?
What does this affect?

**Review by VenelinMartinov — APPROVED** [review]

File: pkg/tfbridge/provider.go (hunk 8 above)
Comment (t0yv0) [reply pkg/tfbridge/provider.go:1332]: It's a bug introduced recently in https://github.com/pulumi/pulumi-terraform-bridge/commit/39a92a469cf6ec3452fdd446d26119852909d6af specifically to unlock this PR and previously unused/untested.

Comment (VenelinMartinov) [reply pkg/tfbridge/provider.go:1332]: Looks like a refactor to me wrt the `t` parameter - https://github.com/pulumi/pulumi-terraform-bridge/commit/39a92a469cf6ec3452fdd446d26119852909d6af#diff-1dee36d7f4752b488fc6d27ff19bd5e68277063de183d75ef29a02b8bf989045R1332

I don't see how that commit introduces a bug there?

File: pkg/tfbridge/provider_test.go [hunk 9]
```diff
@@ -2698,3 +2712,258 @@ func TestPreConfigureCallbackEmitsFailures(t *testing.T) {
...
+		    "id": "res1",
+		    "urn": "urn:pulumi:dev::mystack::ExampleResource::res1name",
+		    "properties": {}
+		  },
+		  "response": {}
+		}`)
+	})
+}
+
+func TestRefresh(t *testing.T) {
```
Comment (VenelinMartinov) [inline pkg/tfbridge/provider_test.go:2812]: These are the first import and refresh tests, right?

File: pkg/tfbridge/provider.go (hunk 8 above)
Comment (t0yv0) [reply pkg/tfbridge/provider.go:1332]: The refactor wasn't tested or used. NewDestroyDiff under default options ignores `t`. The intent though was to pass the Terraform resource name in that parameter, similarly to other methods on shim.Provider. Not the Pulumi type.

File: pkg/tfbridge/provider_test.go (hunk 9 above)
Comment (t0yv0) [reply pkg/tfbridge/provider_test.go:2812]: TestImport tries to emulate what happens during `pulumi import`, and TestRefresh during `pulumi refresh`. Both map to shim.Provider Read method but it's a bit difficult for me to remember which is which. Having explicit in the tests remind me. This makes sure to test both branches of `isRefresh bool` inside Read. Prior to this test, the not found cases were not covered by tests at all.

**Pull request merged**